### PR TITLE
SYNR-1110

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -207,23 +220,67 @@ More information can be found [here](http://www.diveintopython3.net/strings.html
 | pyVersion        | Returns the version of Python                      | `pyVersion()`                                                        |
 
 
-# Usage Examples   
+# Wrapping python packages
+
+While using the original PythonInR to wrap a python package in R, automatically generates R functions and documentation, we added a few helper functions in out version of PythonEmbedInR:
```
Comment (kimyen) [inline README.md]: "out" -> "our"

File: R/PythonPkgWrapperUtils.R [hunk 2]
```diff
@@ -0,0 +1,443 @@
+# ------------------------------------------------------------------------------
+#
+#   Helpers for wrapping python packages
+#
+# ------------------------------------------------------------------------------
+
+#' Determines args and kwargs
+#'
+#' This function determines args and kwargs for a given list of values.
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: This function takes the list of arguments passed to an R function and groups them into the (1) unnamed / positional arguments and the (2) the named / keyword arguments to pass to the corresponding Python function.

File: R/PythonPkgWrapperUtils.R [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+#'
+#' @param ... the input values
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: @param ... the list of arguments passed to an R function

File: R/PythonPkgWrapperUtils.R [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+#' @return A list of args and kwargs
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: The grouping of arguments into 'args' (the unnamed or positional arguments) and 'kwargs' (the named or keyword arguments) to be passed to the corresponding Python function.

File: R/PythonPkgWrapperUtils.R [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+
+#' Clean up stack trace
+#'
+#' This function clean up stack trace for a callable.
+#'
+#' @param callable the callable to be called
+#' @param args the input to be passed to callable
+#' @return the result of calling callable with args after cleaning up the stack trace
+cleanUpStackTrace <- function(callable, args) {
+  conn <- textConnection("outputCapture", open = "w")
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: textConnection() should have the parameter "local=TRUE".  Please see SYNR-1240 and move the associated test from synapser to PythonEmbedInR

File: R/PythonPkgWrapperUtils.R [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        } else {
+          kwargs[[valuenames[[i]]]] <- values[[i]]
+        }
+      }
+    }
+  }
+  list(args = args, kwargs = kwargs)
+}
+
+#' Clean up stack trace
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: The purpose of this function is to remove the Python stack trace from an error message generated when calling Python from R.  This makes the command line response more readable when an error occurs.  To support debugging the stack trace truncation can be overridden by setting the global option 'verbose' to TRUE.

File: R/PythonPkgWrapperUtils.R [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+#'
+#' This function clean up stack trace for a callable.
+#'
+#' @param callable the callable to be called
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: @param callable the function to be called

File: R/PythonPkgWrapperUtils.R [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+#' @param args the input to be passed to callable
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: @param args the arguments to be passed to the function 'callable'

File: R/PythonPkgWrapperUtils.R [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+#' @return the result of calling callable with args after cleaning up the stack trace
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: @return the result of calling the given function with the given arguments

File: R/PythonPkgWrapperUtils.R [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+# ------------------------------------------------------------------------------
+
+# This is factored out of autoGenerateRdFiles so it can be called during testing
+initAutoGenerateRdFiles<-function(templateDir) {
+  dictDocString<<-getDictDocString(templateDir)	
+}
+
+#' Create the autogenerated files
+#'
+#' This function generate R docs from python docs and saves the doc files to the target dir <srcRootDir>/auto-man
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: generate -> generates

File: R/PythonPkgWrapperUtils.R [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+#'
+#' @param srcRootDir is the root directory for the code base (i.e., prior to installation)
+#' @param functionInfo list of function to generate docs
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: function -> functions

File: R/PythonPkgWrapperUtils.R [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+#' @param classInfo list of class to generate docs
```
Comment (brucehoff) [inline R/PythonPkgWrapperUtils.R]: list of classes for which to generate doc's

File: man/cleanUpStackTrace.Rd [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+\arguments{
+\item{callable}{the callable to be called}
+
+\item{args}{the input to be passed to callable}
+}
+\value{
+the result of calling callable with args after cleaning up the stack trace
+}
+\description{
+This function clean up stack trace for a callable.
```
Comment (brucehoff) [inline man/cleanUpStackTrace.Rd]: clean->cleans

File: README.md (hunk 1 above)
Comment (brucehoff) [reply README.md]: I would just say:

"The following tools help generate R functions wrapping Python functions along with reference documentation (.Rd files) for the wrapper functions, the documentation being generated from the Sphinx-based Python doc-strings."

File: README.md [hunk 14]
```diff
@@ -207,23 +220,67 @@ More information can be found [here](http://www.diveintopython3.net/strings.html
...
+
+## Determine args and kwargs (determineArgsAndKwArgs)
```
Comment (brucehoff) [inline README.md]: determineArgsAndKwArgs() is really just a small utility which is part of the larger framework for generating wrapper functions.  Do you think it makes more sense to add the larger framework here?  determineArgsAndKwArgs() would then just be an internal method and there'd be no need for extensive documentation.

File: README.md [hunk 15]
```diff
@@ -207,23 +220,67 @@ More information can be found [here](http://www.diveintopython3.net/strings.html
...
+
+There are many ways to wrap a Python function in R. To create an R wrapper on top of the Python function without making any changes in the function signature, we need to figure out the user's input and pass them along to the Python function. This helper function is used to determine the parameters to pass to the underlying Python function.
+
+| example                                             | output                                                      |
+| `determineArgsAndKwArgs()`                          | `list(args=list(), kwargs=list())`                          |
+| `determineArgsAndKwArgs("foo", "bar")`              | `list(args=list("foo", "bar"), kwargs=list())`              |
+| `determineArgsAndKwArgs("foo", "bar", x="baz")`     | `list(args=list("foo", "bar"), kwargs=list(x="baz"))`       |
+| `determineArgsAndKwArgs(a="foo", b="bar", c="baz")` | `list(args=list(), kwargs=list(a="foo", b="bar", c="baz"))` |
+
+## Clean up stack trace (cleanUpStackTrace)
```
Comment (brucehoff) [inline README.md]: cleanUpStackTrace() is really just a small utility which is part of the larger framework for generating wrapper functions. Do you think it makes more sense to add the larger framework here? cleanUpStackTrace() would then just be an internal method and there'd be no need for extensive documentation.

File: R/PythonPkgWrapperUtils.R (hunk 11 above)
Comment (kimyen) [reply R/PythonPkgWrapperUtils.R]: Is the example in README.md file for this function helpful?

File: README.md (hunk 14 above)
Comment (kimyen) [reply README.md]: Note that we didn't take synapseclient and wrap every functions & classes in that python package. We cherry pick functions and classes under synapseclient.Synapse (+ synapseclient.table.Table) for synapser package. We then cherry pick synapseutils for synapserutils.

**Comment by kimyen** [comment]
Thanks @brucehoff! I will make the suggested changes. And I will think more about what we can / should provide in PythonEmbedInR.

**Pull request closed without merging**